### PR TITLE
Add cblas to libs when compiling without MKL

### PR DESCRIPTION
**Pull request by kpu**
Otherwise on gentoo and CSD3 (without MKL module loaded), I get
missing cblas_gemm link errors

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -271,7 +271,7 @@ if(COMPILE_CPU)
     find_package(BLAS)
     if(BLAS_FOUND)
       include_directories(${BLAS_INCLUDE_DIR})
-      set(EXT_LIBS ${EXT_LIBS} ${BLAS_LIBRARIES})
+      set(EXT_LIBS ${EXT_LIBS} ${BLAS_LIBRARIES} cblas)
```
Comment (emjotde) [inline CMakeLists.txt]: This should rather be done by the typical Find* semantics of CMake. I don't think is clear at all that libcblas exists at this point?.

Comment (emjotde) [reply CMakeLists.txt]: Shouldn't "cblas" appear inside ${BLAS_LIBRARIES} if present?

Comment (kpu) [reply CMakeLists.txt]: I guess FindBLAS finds the fortran library, not the C library:
```
# This module finds an installed fortran library that implements the
# BLAS linear-algebra interface (see http://www.netlib.org/blas/).  The
# list of libraries searched for is taken from the autoconf macro file,
# acx_blas.m4 (distributed at
# http://ac-archive.sourceforge.net/ac-archive/acx_blas.html).
```
So I guess to be pedantic, it's within its rights to not provide the cblas library for C. 

cblas is part of OpenBLAS: https://github.com/xianyi/OpenBLAS/issues/1297

CMake doesn't have an included FindCBLAS or FindOpenBLAS but both of these are floating around other people's projects.

Comment (emjotde) [reply CMakeLists.txt]: So, are we doing anything here? The current solution does fine OpenBLAS for me, at least on Ubuntu.

**Comment by XapaJIaMnu** [comment]
Added a FInd_CBLAS file based on https://github.com/Eyescale/CMake/blob/master/FindCBLAS.cmake that will check for a library containing the `cblas_sgemm` symbol in the following order: openblas, cblas, lapacke. If `cblas_sgemm` is contained in openblas, the behaviour is the same as in master, except we will have two separate `/usr/lib/openblas.so` in the build commands. On the other hand if openblas does not contain `cblas_sgemm` we will add a library containing it to the link path.

I also check for the include paths and add them to the include path. It should check for openblas first and only if openblas doesn't server our purpose, it will move to cblas.

This should also help with mac compilation

**Comment by XapaJIaMnu** [comment]
@emjotde , are we using openblas for anything else but the `cblass_sgemm` calls? If so, we can just supersede find openblas with finding CBLAS. Find openblas is also a bit buggy on mac...

**Comment by XapaJIaMnu** [comment]
Bump

**Comment by kpu** [comment]
This is still broken for me...

**Comment by emjotde** [comment]
Testing this now. With a student base-sized model (512/2048) with 6-encoder layers, 2-decoder layers, SSRU on an AVX2 CPU (Intel(R) Xeon(R) CPU E5-2620 v4 @ 2.10GHz), time to translate first 100 sentences from newstest2014:

```
Openblas: 28.2s
MKL: 6.9s
FBGEMM: 4.2s
```

Before anyone blames Marian to not reach the numbers from the paper :)

**Comment by emjotde** [comment]
Feels to me like we should put a big, fat warning when someone compiles with Openblas.

**Comment by emjotde** [comment]
@snukky This is another example that we should output compilation information with each run, in this case what blas libraries are going to be used in the default case.

**Comment by alvations** [comment]
It seems like this `sgemm` out of scope re-surfaced again in 3aed9143

```
$ cmake .. -DUSE_SENTENCEPIECE=ON -DUSE_FBGEMM=ON -DCOMPILE_CPU=ON
-- The CXX compiler identification is GNU 7.5.0
-- The C compiler identification is GNU 7.5.0
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Project name: marian
-- Project version: v1.9.33;+3aed9143
CMake Warning at CMakeLists.txt:57 (message):
  CMAKE_BUILD_TYPE not set; setting to Release


-- Checking support for CPU intrinsics
-- Could not find hardware support for AVX512 on this machine.
-- SSE2 support found
-- SSE3 support found
-- SSE4.1 support found
-- SSE4.2 support found
-- AVX support found
-- AVX2 support found
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
-- Found CUDA: /usr/local/cuda (found suitable version "10.2", minimum required is "9.0") 
-- Found CUDA libraries: /usr/local/cuda/lib64/libcurand.so /usr/local/cuda/lib64/libcusparse.so /usr/lib/x86_64-linux-gnu/libcublas.so
-- Found Tcmalloc: /usr/lib/x86_64-linux-gnu/libtcmalloc_minimal.so
-- Could NOT find MKL (missing: MKL_LIBRARIES MKL_INCLUDE_DIRS MKL_INTERFACE_LIBRARY MKL_SEQUENTIAL_LAYER_LIBRARY MKL_CORE_LIBRARY) 
-- Looking for sgemm_
-- Looking for sgemm_ - not found
-- Looking for sgemm_
-- Looking for sgemm_ - found
-- Found BLAS: /usr/local/lib/libmkl_intel_lp64.so;/usr/local/lib/libmkl_intel_thread.so;/usr/local/lib/libmkl_core.so;/usr/local/lib/libiomp5.so;-lpthread;-lm;-ldl  
-- Checking for [openblas]
-- Checking for [openblas] -- includes found
-- Checking for [cblas]
-- Checking for [cblas] -- includes found
-- Checking for [lapacke]
-- Checking for [lapacke] -- includes found
-- Looking for cblas_sgemm
-- Looking for cblas_sgemm - not found
-- CBLAS library not found. Please specify library location
-- Performing Test COMPILER_SUPPORTS_AVX512
-- Performing Test COMPILER_SUPPORTS_AVX512 - Success
-- Found OpenMP_C: -fopenmp (found version "4.5") 
-- Found OpenMP_CXX: -fopenmp (found version "4.5") 
-- Found OpenMP: TRUE (found version "4.5")  
-- Performing Test __CxxFlag__fmerge_all_constants
-- Performing Test __CxxFlag__fmerge_all_constants - Success
** AsmJit Summary **
   ASMJIT_DIR=/ssd1/lib/marian-dev/src/3rd_party/fbgemm/third_party/asmjit
   ASMJIT_TEST=FALSE
   ASMJIT_TARGET_TYPE=STATIC
   ASMJIT_DEPS=pthread;rt
   ASMJIT_LIBS=asmjit;pthread;rt
   ASMJIT_CFLAGS=-DASMJIT_STATIC
   ASMJIT_PRIVATE_CFLAGS=-Wall;-Wextra;-fno-math-errno;-fno-threadsafe-statics;-DASMJIT_STATIC
   ASMJIT_PRIVATE_CFLAGS_DBG=
   ASMJIT_PRIVATE_CFLAGS_REL=-O2;-fmerge-all-constants
-- Check if compiler accepts -pthread
-- Check if compiler accepts -pthread - yes
   ASMJIT COMPILE FLAGS: -fno-math-errno;-fno-threadsafe-statics;-DASMJIT_STATIC;$<$<CONFIG:Debug>:>;$<$<NOT:$<CONFIG:Debug>>:-O2;-fmerge-all-constants>
-- VERSION: 0.1.6
-- Found Protobuf: /usr/lib/x86_64-linux-gnu/libprotobuf.so;-pthread (found version "3.0.0") 
-- Found TCMalloc: /usr/lib/x86_64-linux-gnu/libtcmalloc_minimal.so
-- Found Doxygen: /usr/bin/doxygen (found version "1.8.13") found components:  doxygen missing components:  dot
-- Configuring done
-- Generating done
-- Build files have been written to: /ssd1/lib/marian-dev/build
```

Then 

```
$ make -j $(nproc)
...
[ 72%] Building NVCC (Device) object src/CMakeFiles/marian_cuda.dir/translator/marian_cuda_generated_nth_element.cu.o
Scanning dependencies of target marian
[ 72%] Building CXX object src/CMakeFiles/marian.dir/common/logging.cpp.o
[ 72%] Building CXX object src/CMakeFiles/marian.dir/common/cli_wrapper.cpp.o
[ 73%] Building CXX object src/CMakeFiles/marian.dir/common/aliases.cpp.o
[ 74%] Building CXX object src/CMakeFiles/marian.dir/common/config_validator.cpp.o
[ 74%] Building CXX object src/CMakeFiles/marian.dir/common/utils.cpp.o
[ 74%] Building CXX object src/CMakeFiles/marian.dir/common/options.cpp.o
[ 74%] Building CXX object src/CMakeFiles/marian.dir/common/fastopt.cpp.o
[ 74%] Building CXX object src/CMakeFiles/marian.dir/common/version.cpp.o
[ 75%] Building CXX object src/CMakeFiles/marian.dir/common/binary.cpp.o
[ 75%] Building CXX object src/CMakeFiles/marian.dir/common/cli_helper.cpp.o
[ 75%] Building CXX object src/CMakeFiles/marian.dir/common/config_parser.cpp.o
[ 76%] Building CXX object src/CMakeFiles/marian.dir/common/config.cpp.o
[ 76%] Building CXX object src/CMakeFiles/marian.dir/common/file_utils.cpp.o
[ 76%] Building CXX object src/CMakeFiles/marian.dir/common/io.cpp.o
[ 76%] Building CXX object src/CMakeFiles/marian.dir/common/types.cpp.o
[ 77%] Building CXX object src/CMakeFiles/marian.dir/common/filesystem.cpp.o
[ 77%] Building CXX object src/CMakeFiles/marian.dir/data/sentencepiece_vocab.cpp.o
[ 78%] Building CXX object src/CMakeFiles/marian.dir/data/alignment.cpp.o
...
[ 92%] Building CXX object src/CMakeFiles/marian.dir/training/communicator.cpp.o
[ 93%] Building CXX object src/CMakeFiles/marian.dir/training/scheduler.cpp.o
[ 93%] Building CXX object src/CMakeFiles/marian.dir/microsoft/quicksand.cpp.o
[ 93%] Building CXX object src/CMakeFiles/marian.dir/layers/lsh.cpp.o

/ssd1/lib/marian-dev/src/layers/lsh.cpp: In lambda function:
/ssd1/lib/marian-dev/src/layers/lsh.cpp:84:9: error: ‘sgemm’ was not declared in this scope
         sgemm(false, true, 1, 1, dimIn, 1.0f, currQueryPtr, dimIn, currWPtr, dimIn, 1.0f, &currOutPtr[relPos], 1);
         ^~~~~
At global scope:
cc1plus: error: unrecognized command line option ‘-Wno-unknown-warning-option’ [-Werror]
cc1plus: all warnings being treated as errors
src/CMakeFiles/marian.dir/build.make:1050: recipe for target 'src/CMakeFiles/marian.dir/layers/lsh.cpp.o' failed
make[2]: *** [src/CMakeFiles/marian.dir/layers/lsh.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
```

**Comment by emjotde** [comment]
@alvations Can you open an issue for that please?

**Comment by alvations** [comment]
Okay =)

----

BTW, instaling the `libopenblas-dev` will automatically link the sgemm back but I think that's not the expected behavior, when we don't want blas dependency.

**Pull request merged**